### PR TITLE
feat(siit): Initial version

### PR DESCRIPTION
**Pull request by micheelengronne**
### 1. Why is this pull request needed and what does it do?

This PR provides support for:

* [RFC 6052](https://tools.ietf.org/html/rfc6052), the SIIT mechanism
* [RFC 7757](https://tools.ietf.org/html/rfc7757), the explicit address mappings (eam) mechanism

With this new plugin a system can use coredns to serve correct records for services in ipv6-only to ipv4-only client via the help of SIIT (of course the network still needs to do the conversion with a tool like jool for instance).

### 2. Which issues (if any) are related?

* https://github.com/coredns/coredns/issues/8084

### 3. Which documentation changes (if any) need to be made?

None needed. Except the README of the plugin.

### 4. Does this introduce a backward incompatible change or deprecation?

No.

### CODE REVIEW

**Comment by micheelengronne** [comment]
Hi @bradbeam can you retrigger the pipeline please ?

**Comment by micheelengronne** [comment]
Hi @yongtang what is the procedure to have my PR reviewed and merged please ?

**Comment by yongtang** [comment]
@micheelengronne Thanks for the PR. The PR is fairly large so it may take a little time. I am quite occupied this week but will try to take a look during the weekend.

**Comment by micheelengronne** [comment]
Thank you @yongtang

**Comment by micheelengronne** [comment]
Hi @yongtang did you have time to take a look ? Do you need help ? I intend to improve a bit the README and add one feature but after your first feedbacks.

**Comment by micheelengronne** [comment]
I simplified and improved the README.

File: plugin.cfg [hunk 1]
```diff
@@ -76,5 +76,6 @@ erratic:erratic
 whoami:whoami
 on:github.com/coredns/caddy/onevent
 sign:sign
+siit:siit
```
Comment (houyuwushang) [inline plugin.cfg]: **Blocking:** `siit` needs to run before the response-producing plugins whose replies it intercepts, likely adjacent to `dns64` and before `cache`/the backends. CoreDNS builds the chain in `plugin.cfg` order, regardless of textual Corefile order. At this position, `forward`, `file`, `kubernetes`, etc. are above `siit` and normally return without calling it.

I reproduced this with a real CoreDNS server configured with `siit` and `forward`: the local upstream returned A NODATA and `AAAA 64:ff9b::192.0.2.42`, but the client still received an empty A answer. Please move the directive and add a full-Corefile integration test so ordering cannot regress.

File: plugin/siit/siit.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	for _, rr := range resp.Answer {
+		header := rr.Header()
+		// 5.3.3: All other RR's MUST be returned unchanged
+		if header.Rrtype != dns.TypeAAAA {
+			ret.Answer = append(ret.Answer, rr)
+			continue
+		}
+
+		if header.Rrtype == dns.TypeAAAA {
+			a, _ := to4(d.Eam4, d.IPv6Prefix, rr.(*dns.AAAA).AAAA)
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: **Blocking:** please handle the result of `to4` before constructing the A RR. `to4` returns `(nil, nil)` when an AAAA address matches neither the EAM table nor `ipv6_prefix`; this code then appends an A RR with empty RDATA (wire `RDLENGTH=0`). A response containing an unrelated AAAA such as `2001:db8::1` reproduces it.

Only append an A for a successful mapping. If no AAAA is mappable, return the appropriate empty/error response instead. Please test unmapped-only and mixed mapped/unmapped answers.

File: plugin/siit/siit.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+// to4 takes an IPv6 address and an eam and returns an IPv4 address.
+func to4(eam map[string]net.IP, ipv6prefix *net.IPNet, addr net.IP) (net.IP, error) {
+	addr = addr.To16()
+	if addr == nil || addr.To4() != nil {
+		return nil, errors.New("not a valid IPv6 address")
+	}
+
+	if ipv6prefix.Contains(addr) {
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: **Blocking:** the lookup order is reversed. [RFC 7757 section 3.3.2](https://www.rfc-editor.org/rfc/rfc7757#section-3.3.2) requires searching the EAMT first and falling back to RFC 6052 algorithmic translation only when no EAM matches. With `ipv6_prefix 64:ff9b::/96` and an explicit mapping `203.0.113.9 <-> 64:ff9b::192.0.2.1`, this currently returns `192.0.2.1` and silently ignores the EAM. Please reverse the checks and add a precedence regression test.

File: plugin/siit/setup.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Test for valid prefix
+	n, total := pref.Mask.Size()
+	if total != 128 {
+		return nil, c.Errf("invalid netmask %d IPv6 address: %q", total, pref)
+	}
+	if n%8 != 0 || n < 32 || n > 96 {
```
Comment (houyuwushang) [inline plugin/siit/setup.go]: **Blocking:** byte alignment is not sufficient for an RFC 6052 prefix. [RFC 6052 section 2.2](https://www.rfc-editor.org/rfc/rfc6052#section-2.2) permits exactly /32, /40, /48, /56, /64, and /96; this condition also accepts /72, /80, and /88. Please validate the explicit set and test all allowed lengths plus rejected gaps.

For lengths below /96, also verify the reserved `u` octet before treating an address as RFC 6052-encoded; the current extractor skips a nonzero value and still produces an IPv4 address.

File: plugin/siit/setup.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		return nil, c.Errf("invalid prefix length %q", pref)
+	}
+
+	return pref, nil
+}
+
+func parseEam(c *caddy.Controller) (map[int]net.IP, error) {
+	args := c.RemainingArgs()
```
Comment (houyuwushang) [inline plugin/siit/setup.go:111]: **Blocking:** validate `len(args) == 2` before indexing. A Corefile containing `eam 192.0.2.1` currently panics with `index out of range [1] with length 1` instead of returning a configuration error, while extra arguments are silently ignored. Please return `c.ArgErr()` for either arity mismatch and add malformed-directive tests.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
Thanks for working on #8084. I reviewed commit `651da1594f` against the current plugin chain and the RFC 6052/RFC 7757 mapping rules.

Local verification:
- `go test ./plugin/siit -run "^(TestSIIT|TestSetupSiit)$" -count=20`
- `go test -race ./plugin/siit -run "^(TestSIIT|TestSetupSiit)$" -count=5`
- `go vet ./plugin/siit`
- `go test ./core/dnsserver ./core/plugin`
- a full CoreDNS + local upstream integration test

The existing tests, race test, vet, and registration tests pass. The integration test and focused negative tests expose the blocking issues below. In particular, the current directive position prevents the plugin from running in a normal `forward`/authoritative setup, and several mapping paths can return invalid or incorrect A answers.

Please also cover these response-contract and scope points before re-review:

1. A SERVFAIL/REFUSED/NXDOMAIN from the second AAAA lookup is currently reset to NOERROR by `SetReply`. Synthesis should happen only when that lookup provides usable mapped AAAA data; otherwise preserve the appropriate empty/error response and header flags.
2. This implements DNS46 record synthesis for a controlled SIIT/SIIT-DC deployment, not packet-level SIIT itself. RFC 6145 is obsolete, and RFC 6052 defines the address embedding. Please update the package/README references and deployment scope.
3. The current `eam` syntax implements only host mappings (/32 to /128), whereas RFC 7757 defines prefix EAMs and longest-prefix matching. Either document this subset explicitly or implement the advertised prefix behavior.
4. Define and test DNSSEC behavior. Copying a RRSIG that covers AAAA while replacing the AAAA RRset with A produces a contradictory response; DO/CD/AD handling or an explicit guarded limitation is needed.

Please add regression tests for each inline case, including a real Corefile ordering test.

**Comment by micheelengronne** [comment]
Thank you @houyuwushang I did all the changes requested except the Corefile ordering test (or maybe I misunderstood this one).

File: plugin/siit/siit.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp == nil || resp.Rcode != dns.RcodeSuccess {
+		// Not a transport error — we got an answer, it's just SERVFAIL/
+		// REFUSED/NXDOMAIN/etc. Don't synthesize from it; return the
+		// original response untouched so ServeDNS writes that instead.
+		return origResponse, nil
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: **Blocking:** this still hides an error from the second lookup. With an initial A NODATA/NOERROR response and a secondary AAAA SERVFAIL, `DoSIIT` returns the original NOERROR response. The new `TestDoSIITPassesThroughNegativeResponse` and NXDOMAIN test encode that behavior, rather than preserving the secondary result.

For the reverse-synthesis workflow modeled here, [RFC 6147 section 5.1.6](https://www.rfc-editor.org/rfc/rfc6147#section-5.1.6) says the second query's empty result or error is the basis for the client response. Please preserve that response's RCODE/flags/authority/additional data while restoring the original A question and request ID, and change the tests to expect the secondary error instead of NOERROR.

File: plugin/siit/siit.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+			ret.Answer = append(ret.Answer, rr)
+			continue
+		}
+
+		if header.Rrtype == dns.TypeAAAA {
+			a, mapped := to4(d.Eam4, d.IPv6Prefix, rr.(*dns.AAAA).AAAA)
+			if !mapped {
+				// No EAM entry and address isn't in ipv6_prefix — nothing to synthesize
+				// for this record; skip it rather than emitting an empty-RDATA A record.
+				continue
```
Comment (houyuwushang) [inline plugin/siit/siit.go:246]: **Blocking:** skipping an unmappable AAAA prevents the malformed wire RR, but the unmapped-only case still returns the newly assembled empty `ret`. That loses the original A-NODATA response's SOA and header flags. I reproduced this with an authoritative A NODATA containing an SOA and a successful AAAA answer containing only `2001:db8::1`; the final response has `AA=false` and no authority section.

Please track whether at least one A was synthesized. If none was synthesized, return the appropriate original empty response; if the response is mixed, retain only the successfully mapped records. Add `Synthesize`/`ServeDNS` regressions for both unmapped-only and mixed mapped/unmapped answers.

File: plugin/siit/siit.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	}
+
+	// RFC 7757 §3.3.2: search the EAM table first.
+	if eam[addr.String()] != nil {
+		v4 := eam[addr.String()]
+		return v4, true
+	}
+
+	// Fall back to RFC 6052 algorithmic translation only if no EAM matched.
+	if ipv6prefix.Contains(addr) {
```
Comment (houyuwushang) [inline plugin/siit/siit.go:312]: **Blocking:** the reserved `u` octet is still not validated on the address being translated. For prefix `2001:db8:122:344::/64`, `2001:db8:122:344:ffc0:2:2100:0` has a nonzero byte at bits 64-71, but this path currently maps it to `192.0.2.33`. [RFC 6052 section 2.2](https://www.rfc-editor.org/rfc/rfc6052#section-2.2) fixes `u` at bits 64-71 (byte index 8) for every format.

The new setup switch checks byte `n/8` in the textual prefix for /32 through /64, which is the first host/embedded-IPv4 byte rather than `u`; it also accepts a /96 prefix such as `2001:db8:122:344:ff00::/96`, whose bits 64-71 are nonzero. Keep EAM lookup first, but reject algorithmic translation when `addr[8] != 0` (and preferably reject an invalid /96 configuration early). Please test the six RFC 6052 Table 1 vectors plus a nonzero-u address.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
Thanks for the quick update. I re-reviewed `fab3614f` and confirmed that this head fixes the directive position, EAM-before-algorithmic precedence, exact allowed prefix lengths, malformed `eam` arity, and the malformed empty-RDATA A record. A real `siit` + `forward` CoreDNS server now synthesizes `A 192.0.2.42` from `AAAA 64:ff9b::192.0.2.42`, so the ordering change is effective.

I can still reproduce three blocking address/response cases, described inline.

The full Corefile regression test is still needed because the package tests instantiate `SIIT` directly and therefore cannot detect a future `plugin.cfg` ordering regression. A test under `test/` can start a local `dnstest.NewServer`, return A NODATA plus a mapped AAAA, start CoreDNS with `siit` and `forward`, then assert that an A query returns the mapped address.

Please also finish the earlier scope/documentation item: the package comment still says this performs SIIT and cites obsolete RFC 6145; the README still says `eam` works vice versa/arbitrarily although this implementation performs exact IPv6-host-to-IPv4 mappings only, and its relative `dns64` link is broken.

Local verification on this head:
- `go test ./plugin/siit -count=20`
- `go test -race ./plugin/siit -count=5`
- `go vet ./plugin/siit`
- `go test ./core/dnsserver ./core/plugin -count=1`
- `go test ./test -run '^TestReadme$' -count=1`
- temporary full-Corefile `siit` + `forward` integration regression: passes
- RFC 6052 Table 1 vectors for /32, /40, /48, /56, /64, /96: pass
- focused nonzero-u, second-query-SERVFAIL, and unmapped-only negative-response regressions: fail as described below

DCO is green; the CircleCI `kubernetes-tests` context is still pending/not reported.

**Comment by micheelengronne** [comment]
I added the integration test and fixed the 3 issues @houyuwushang

**Comment by micheelengronne** [comment]
I fixed the linting.

**Comment by micheelengronne** [comment]
@houyuwushang @yongtang can you retrigger please ?

File: plugin/siit/siit.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+	return true
+}
+
+// DoSIIT takes an (empty) response to an A question, issues the AAAA request,
+// and synthesizes the answer. Returns the response message, or error on internal failure.
+func (d *SIIT) DoSIIT(ctx context.Context, w dns.ResponseWriter, r *dns.Msg, origResponse *dns.Msg) (*dns.Msg, error) {
+	req := request.Request{W: w, Req: r}
+	defaultreq := dns.TypeAAAA
+
+	resp, err := d.Upstream.Lookup(ctx, req, req.Name(), defaultreq)
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: **Blocking:** this full-server lookup can make `siit` and `dns64` recursively synthesize each other's internal queries. I reproduced it with the actual handler order (`dns64 -> siit -> empty backend`) and an IPv6 client: the A query enters `siit`, its internal AAAA query enters `dns64`, and `dns64`'s internal A query enters `siit` again. I stopped the test after eight internal lookups; a real server has no bound here. The same occurs for IPv4 clients when `dns64 allow_ipv4` is enabled.

Please mark the synthesis lookup in the context and make nested `siit` A handling bypass translation (or otherwise prevent the two plugins from re-entering each other), and commit a combined-plugin regression. Both plugins are built in and adjacent in `plugin.cfg`, so this configuration must not be able to exhaust the stack.

File: plugin/siit/siit.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp == nil || resp.Rcode != dns.RcodeSuccess {
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: **Blocking:** the secondary response is not yet converted back into the original client transaction. `upstream.Lookup` changes the question to AAAA, so returning `resp` directly sends an AAAA question in the response to the client's A query. My focused reproduction preserved SERVFAIL/AA/RA/SOA but received `Question[0].Qtype == AAAA` instead of A. RFC 6147 section 5.4 requires the original initiator question in the assembled response.

There is also a nil path here: when the internal lookup produces `(nil, nil)`, `DoSIIT` returns `(nil, nil)` and `ServeDNS` passes it to `WriteMsg` and then dereferences `msg.Rcode`.

Please copy the secondary error response, restore the original ID/question while retaining its RCODE, flags, authority, and additional sections, and handle a nil response as an error or a defined fallback. Add ServeDNS-level tests for SERVFAIL/NXDOMAIN and nil.

File: plugin/siit/siit.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	out := d.Synthesize(r, origResponse, resp)
+	return out, nil
+}
+
+// Synthesize merges the AAAA response and the records from the A response
+func (d *SIIT) Synthesize(origReq, origResponse, resp *dns.Msg) *dns.Msg {
+	ret := dns.Msg{}
+	ret.SetReply(origReq)
```
Comment (houyuwushang) [inline plugin/siit/siit.go:206]: **Blocking:** `SetReply(origReq)` clears the response-side AA and RA flags. With a mapped AAAA response carrying `AA=true` and `RA=true`, the synthesized A response has both false. In practice, authoritative backends lose AA and recursive/forwarded responses lose RA. RFC 6147 section 5.4 requires header fields to follow the normal authoritative/recursive rules while the question comes from the original query.

Please assemble from the appropriate response header, then restore the original transaction fields and deliberately clear only fields invalidated by synthesis (notably AD). Add authoritative and recursive flag regressions.

File: plugin/siit/siit.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+	// otherwise do the actual SIIT request and response synthesis
+	msg, err := d.DoSIIT(ctx, w, r, nw.Msg)
+	if err != nil {
+		// err means we weren't able to even issue the A or AAAA request
+		// to CoreDNS upstream
+		return dns.RcodeServerFailure, err
+	}
+
+	RequestsTranslatedCount.WithLabelValues(metrics.WithServer(ctx)).Inc()
```
Comment (houyuwushang) [inline plugin/siit/siit.go]: The translated counter is incremented for every non-error return from `DoSIIT`, including a secondary SERVFAIL/NXDOMAIN and the unmapped-only fallback that returns `origResponse`. Those requests did not produce a synthetic A record, so the documented metric overcounts. Please return an explicit synthesis result (or equivalent) and increment only when at least one mapped A was emitted; cover the fallback and error cases.

File: plugin/siit/siit_test.go [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+
+// TestExtractIPv4RFC6052Vectors covers all six valid RFC 6052 Table 1
+// prefix lengths and verifies round-trip extraction.
+func TestExtractIPv4RFC6052Vectors(t *testing.T) {
+	v4 := net.ParseIP("192.0.2.33").To4()
+	base := net.ParseIP("2001:db8::")
+
+	for _, pl := range []int{32, 40, 48, 56, 64, 96} {
+		t.Run(fmt.Sprintf("PL_%d", pl), func(t *testing.T) {
+			addr := embedIPv4(base, pl, v4)
```
Comment (houyuwushang) [inline plugin/siit/siit_test.go]: This is not actually a Table 1 vector test: `embedIPv4` mirrors the same byte-placement algorithm as `extractIPv4`, so an offset error copied into both helpers can still round-trip successfully. Please use the six literal RFC 6052 Table 1 IPv6 addresses and assert that each extracts to `192.0.2.33`. I ran those literal fixtures against this head and they all pass, so this is a durable-test correction rather than a production-code change.

File: plugin/siit/README.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
+# siit
+
+## Name
+
+*siit* - enables A<>AAAA translation support for DNS records based on SIIT (IPv6<>IPv4 translation).
```
Comment (houyuwushang) [inline plugin/siit/README.md]: The public scope is still described as bidirectional `A<>AAAA` / `IPv4<>IPv6`, and line 12 still says arbitrary bidirectional mapping. This implementation intercepts only A queries, synthesizes A from AAAA, and supports exact IPv6-host-to-IPv4 EAM entries. Please make the Name/Description match that one-way, host-mapping contract; the directive bullet alone does not remove the conflicting claim.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
Thanks for addressing the previous round. I re-reviewed `fb5d5df2` and the current merge result. This head fixes the nonzero-u translation, unmapped-only and mixed-answer handling, the real Corefile ordering coverage, the package scope comment, and the README link. I also independently checked all six literal RFC 6052 Table 1 vectors; the implementation maps them correctly.

I found three remaining runtime/protocol blockers plus contract and coverage issues, described inline. The most serious is mutual re-entry when `dns64` and `siit` are enabled together; a normal A query can alternate internal AAAA/A lookups until stack exhaustion. The secondary-error path also still returns an AAAA transaction to an A client and has a `(nil, nil)` dereference path, while successful synthesis drops AA/RA.

The committed focused tests pass repeatedly, including race and the full-Corefile test, and all packages compile on the current merge result. Those tests do not exercise the failing cases below. Please add regressions for each before the next review.

**Comment by micheelengronne** [comment]
@houyuwushang I am ready for your next review.

File: plugin/siit/siit.go [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+
+// responseShouldSIIT returns true if the response indicates we should attempt
+// SIIT rewriting:
+// 1. The response has no valid (RFC 5.1.4) A records (RFC 5.1.1)
+// 2. The response code (RCODE) is not 3 (Name Error) (RFC 5.1.2)
+//
+// Note that requestShouldIntercept must also have been true, so the request
+// is known to be of type A.
+func (d *SIIT) responseShouldSIIT(req *request.Request, origResponse *dns.Msg) bool {
+	ty, _ := response.Typify(origResponse, time.Now().UTC())
```
Comment (houyuwushang) [inline plugin/siit/siit.go:123]: **Blocking:** `TC=1` means the original A result is incomplete, not authoritative NODATA. This path currently returns true when the truncated response has no visible A RR. I reproduced it with an initial `NOERROR/TC=1` A response and a mapped secondary AAAA: the client receives one synthetic A with `TC=0`, so it does not retry over TCP and a real A RR omitted by truncation can be hidden.

[RFC 6147 section 5.4](https://www.rfc-editor.org/rfc/rfc6147#section-5.4) keeps the final result subject to normal DNS truncation rules. Please short-circuit synthesis when `origResponse.Truncated` and return the original response, with a ServeDNS regression. A minimal `if origResponse.Truncated { return false }` guard passed the focused probe under race 100 times.

File: plugin/siit/siit_test.go [hunk 16]
```diff
@@ -0,0 +1,1110 @@
...
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+
+	"github.com/coredns/coredns/plugin/metrics"
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/test"
+	"github.com/coredns/coredns/request"
+	"github.com/prometheus/client_golang/prometheus/testutil"
```
Comment (houyuwushang) [inline plugin/siit/siit_test.go:14]: **Blocking (presubmit):** this Prometheus package is third-party, but it is inside the CoreDNS-internal import group. `go test ./test -run '^TestImportOrdering$'` reports:

`import path for "github.com/prometheus/client_golang/prometheus/testutil" is not of the same type "coredns" in "plugin/siit/siit_test.go"`

Please move it below the blank line alongside `github.com/miekg/dns`. I verified that this removes the only PR-specific import-order error; the remaining protobuf reports reproduce on current `master` with my local Go 1.26 toolchain.

File: plugin/siit/siit_dns64_test.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (u reentrantUpstream) Lookup(ctx context.Context, _ request.Request, name string, typ uint16) (*dns.Msg, error) {
+	*u.calls++
+	if *u.calls > u.maxOK {
+		u.t.Fatalf("recursion exceeded bound: %d internal lookups (siit/dns64 re-entering each other)", *u.calls)
+	}
+	m := new(dns.Msg)
+	m.SetQuestion(name, typ)
+	rec := dnstest.NewRecorder(&test.ResponseWriter{})
```
Comment (houyuwushang) [inline plugin/siit/siit_dns64_test.go:47]: **Coverage (non-blocking for the implementation):** this differs from the real `upstream.Lookup`, which wraps `state.W` and therefore preserves the original client address family. `test.ResponseWriter{}` is IPv4, and `AllowIPv4: true` below forces only the IPv4 + `allow_ipv4` branch, despite the test comment saying IPv6 client.

Please use `dnstest.NewRecorder(state.W)` here and preferably parameterize both an actual `test.ResponseWriter6` with default DNS64 and IPv4 with `AllowIPv4`. A state-preserving IPv6 probe passed 100 iterations on this production fix, so the code appears correct; the committed regression should cover the branch it claims.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
Thanks for addressing the previous round. I re-reviewed `cf62560f` against current `master` and independently verified the recursion, secondary-error envelope, AA/RA/AD, metric, and literal RFC 6052 vector fixes. Those earlier blockers are resolved.

Two blockers remain:

1. A truncated initial A response is treated as NODATA. With `TC=1`, no Answer, and a mapped secondary AAAA, this head returns `NOERROR`, one synthetic A, and `TC=0`. The client will not retry over TCP, so a real A RR omitted by truncation can be hidden. Details inline.
2. CoreDNS presubmit rejects the new `testutil` import grouping. Details inline.

The combined dns64/siit fix also works in my state-preserving IPv6 re-entry probe (100 iterations), but the committed test only covers the IPv4 + `allow_ipv4` path despite its comment. Please strengthen that regression as noted inline.

Verification on the current merge tree:

- `go test -race ./plugin/siit -count=10` PASS
- `go test -race ./test -run '^TestSIITAfterForward$' -count=20` PASS
- `go test ./core/dnsserver ./core/plugin -count=5` PASS
- `go test ./plugin/siit -run '^TestSIITDNS64NoRecursion$' -count=500` PASS
- Minimal TC guard probe under race x100 PASS after a local-only patch
- Import regrouping removes the only PR-specific `TestImportOrdering` failure; the remaining protobuf reports reproduce on `master` with local Go 1.26

The full suite otherwise reached all packages; unrelated localhost/proxy and Windows listener tests failed in this environment.

**Comment by micheelengronne** [comment]
@houyuwushang ready for review.

**Comment by micheelengronne** [comment]
The failing test seems to come from another plugin, right ?

**Comment by houyuwushang** [comment]
Yes. The remaining local `TestImportOrdering` reports I mentioned are in the generated `pb/*.go` files and reproduce on current `master`; they are not caused by this PR. The SIIT-specific `testutil` import-order issue is fixed on `7f40b336`, and the current CI run is fully green. I am finishing the current-head re-review now.

File: plugin/siit/siit_dns64_test.go [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+			rec := dnstest.NewRecorder(tc.rw)
+			rc, err := d.ServeDNS(context.Background(), rec, r)
+			if err != nil {
+				t.Fatalf("ServeDNS returned error: %v", err)
+			}
+			if rc != dns.RcodeSuccess {
+				t.Fatalf("unexpected rcode: %v", rc)
+			}
+			if calls > 2 {
```
Comment (houyuwushang) [inline plugin/siit/siit_dns64_test.go]: Non-blocking test hardening: please make this `calls != 2` and say "exactly 2" below. I temporarily changed `NewRecorder(state.W)` back to a fixed IPv4 writer; the committed test still passed 100 times because the IPv6 case made one lookup and `1 <= 2`. With an exact assertion, the unmodified head passed 500 runs plus 20 race runs, while that mutation failed the IPv6 case with `got 1`. This will ensure the new IPv6/default-dns64 case actually exercises dns64's nested A lookup.

**Review by houyuwushang — APPROVED** [review]
Re-reviewed `7f40b336` against current `master`. The previous truncated-response, SIIT import-order, and client-family coverage issues are addressed, and all earlier production blockers remain resolved.

Verification included the exact head and the current merge tree:

- `go test ./plugin/siit -count=20`
- `go test -race ./plugin/siit -count=10`
- `go test -race ./test -run '^TestSIITAfterForward$' -count=20`
- `go vet ./plugin/siit`
- `go test ./core/dnsserver ./core/plugin -count=5`
- focused TC and dns64/SIIT tests, including reverse mutations

Removing the TC guard makes the new truncation regression fail immediately, and removing the synthesis context guard makes both dns64/SIIT client-family cases exceed the recursion bound. The current `master` merge is clean, and all 14 GitHub, DCO, and CircleCI checks pass.

Approved. I left one non-blocking inline suggestion to make the IPv6 branch coverage durable by asserting exactly two internal lookups. Also, please refresh the PR description before merge: it still cites RFC 6145, while the code and README now describe the narrower RFC 6052/RFC 7757 AAAA-to-A synthesis scope.

**Comment by micheelengronne** [comment]
I pushed the last non-blocking change you asked for @houyuwushang

**Review by houyuwushang — APPROVED** [review]
Re-reviewed `c2d5b85d`. The first nine patches are patch-equivalent to the previously approved head; the only new change makes the dns64/SIIT regression require exactly two internal lookups, addressing the prior coverage suggestion.

The strengthened test passed 500 repetitions and 50 race repetitions. Replacing the state-preserving writer with a fixed IPv4 writer now reliably fails the IPv6 case with `got 1`, confirming that the exact assertion catches the intended coverage regression. The initial Windows plugin failure was an unrelated local test-server allocation failure across several packages; the Windows and Ubuntu plugin reruns both passed, and all checks are now green.

Approved.

**Comment by micheelengronne** [comment]
Thank you @houyuwushang

@yongtang I think it is good to merge :)
